File: mycelix-desci/sdk/python/src/mycelix/api/claims.py
# Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
# Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
"""
Claims API
"""

from typing import List, Optional, TYPE_CHECKING
from uuid import UUID

from ..models import (
    Claim,
    CreateClaimRequest,
    AddVerificationRequest,
    AddProvenanceRequest,
    EpistemicTier,
    ClaimContent,
)

if TYPE_CHECKING:
    from ..client import MycelixClient, AsyncMycelixClient


class ClaimsAPI:
    """Synchronous Claims API"""

    def __init__(self, client: "MycelixClient"):
        self.client = client

    def create(
        self,
        tier: EpistemicTier,
        content: ClaimContent,
        creator: str,
    ) -> Claim:
        """
        Create a new claim.

        Args:
            tier: Epistemic tier
            content: Claim content
            creator: Creator identifier

        Returns:
            Created claim

        Example:
            >>> claim = client.claims.create(
            ...     tier=EpistemicTier.E0,
            ...     content=ClaimContent(
            ...         dataset_hash="blake3:abc123",
            ...         description="Novel finding",
            ...         category="longevity",
            ...         keywords=["NAD+"]
            ...     ),
            ...     creator="researcher@uni.edu"
            ... )
        """
        request = CreateClaimRequest(tier=tier, content=content, creator=creator)
        data = self.client._request(
            "POST",
            "/api/v1/claims",
            json=request.model_dump(mode="json"),
        )
        return Claim(**data)

    def get(self, claim_id: UUID) -> Claim:
        """
        Get claim by ID.

        Args:
            claim_id: Claim UUID

        Returns:
            Claim

        Raises:
            ClaimNotFoundError: If claim not found
        """
        data = self.client._request("GET", f"/api/v1/claims/{claim_id}")
        return Claim(**data)

    def list(self, limit: int = 100, offset: int = 0) -> List[Claim]:
        """
        List claims with pagination.

        Args:
            limit: Maximum claims to return
            offset: Number of claims to skip

        Returns:
            List of claims
        """
        data = self.client._request(
            "GET",
            "/api/v1/claims",
            params={"limit": limit, "offset": offset},
        )
        return [Claim(**claim) for claim in data.get("claims", [])]

    def add_verification(
        self,
        claim_id: UUID,
        verifier: str,
        signature: str,
        comments: Optional[str] = None,
    ) -> Claim:
        """Add verification to claim"""
        request = AddVerificationRequest(
            verifier=verifier, signature=signature, comments=comments
        )
        data = self.client._request(
            "POST",
            f"/api/v1/claims/{claim_id}/verifications",
            json=request.model_dump(mode="json"),
        )
        return Claim(**data)

    def add_provenance(
        self,
        claim_id: UUID,
        source: str,
        relationship: str,
    ) -> Claim:
        """Add provenance to claim"""
        request = AddProvenanceRequest(source=source, relationship=relationship)
        data = self.client._request(
            "POST",
            f"/api/v1/claims/{claim_id}/provenance",
            json=request.model_dump(mode="json"),
        )
        return Claim(**data)

    def delete(self, claim_id: UUID) -> None:
        """Delete claim"""
        self.client._request("DELETE", f"/api/v1/claims/{claim_id}")


class AsyncClaimsAPI:
    """Asynchronous Claims API"""

    def __init__(self, client: "AsyncMycelixClient"):
        self.client = client

    async def create(
        self,
        tier: EpistemicTier,
        content: ClaimContent,
        creator: str,
    ) -> Claim:
        """Create a new claim (async)"""
        request = CreateClaimRequest(tier=tier, content=content, creator=creator)
        data = await self.client._request(
            "POST",
            "/api/v1/claims",
            json=request.model_dump(mode="json"),
        )
        return Claim(**data)

    async def get(self, claim_id: UUID) -> Claim:
        """Get claim by ID (async)"""
        data = await self.client._request("GET", f"/api/v1/claims/{claim_id}")
        return Claim(**data)

    async def list(self, limit: int = 100, offset: int = 0) -> List[Claim]:
        """List claims (async)"""
        data = await self.client._request(
            "GET",
            "/api/v1/claims",
            params={"limit": limit, "offset": offset},
        )
        return [Claim(**claim) for claim in data.get("claims", [])]

    async def add_verification(
        self,
        claim_id: UUID,
        verifier: str,
        signature: str,
        comments: Optional[str] = None,
    ) -> Claim:
        """Add verification (async)"""
        request = AddVerificationRequest(
            verifier=verifier, signature=signature, comments=comments
        )
        data = await self.client._request(
            "POST",
            f"/api/v1/claims/{claim_id}/verifications",
            json=request.model_dump(mode="json"),
        )
        return Claim(**data)

    async def add_provenance(
        self,
        claim_id: UUID,
        source: str,
        relationship: str,
    ) -> Claim:
        """Add provenance (async)"""
        request = AddProvenanceRequest(source=source, relationship=relationship)
        data = await self.client._request(
            "POST",
            f"/api/v1/claims/{claim_id}/provenance",
            json=request.model_dump(mode="json"),
        )
        return Claim(**data)

    async def delete(self, claim_id: UUID) -> None:
        """Delete claim (async)"""
        await self.client._request("DELETE", f"/api/v1/claims/{claim_id}")

File: mycelix-workspace/mycelix-desci/src/api/src/handlers/claims.rs
// Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
//! Claims API handlers

use axum::{
    Json,
    extract::{Path, State},
    http::StatusCode,
};
use mycelix_desci_core::claims::{ClaimContent, DesciClaim, Provenance, Verification};
use mycelix_desci_core::storage::StorageBackend;
use std::sync::Arc;
use tracing::{info, instrument};
use uuid::Uuid;

use crate::{
    error::{ApiError, Result},
    metrics,
    models::*,
    state::AppState,
};

/// Create a new claim
#[utoipa::path(
    post,
    path = "/api/v1/claims",
    request_body = CreateClaimRequest,
    responses(
        (status = 201, description = "Claim created successfully", body = ClaimResponse),
        (status = 400, description = "Invalid request", body = ApiError),
    ),
    tag = "claims"
)]
#[instrument(skip(state), fields(category = %req.content.category))]
pub async fn create_claim(
    State(state): State<Arc<AppState>>,
    Json(req): Json<CreateClaimRequest>,
) -> Result<(StatusCode, Json<ClaimResponse>)> {
    info!("Creating new claim");

    // Convert request to core types
    let content = ClaimContent {
        dataset_hash: req.content.dataset_hash,
        description: req.content.description,
        category: req.content.category,
        keywords: req.content.keywords,
        storage_ref: req.content.storage_ref,
        reproducibility_score: req.content.reproducibility_score,
        license: req.content.license,
    };

    // Create claim
    let claim = DesciClaim::new(req.tier, content, req.creator);

    // Store claim
    state.storage.store(&claim).await?;

    // Add to query index
    state.query_engine.write().await.add_claim(&claim).await;

    // Track metrics
    metrics::track_claim_creation(&format!("{:?}", claim.epistemic_tier));
    metrics::track_storage_operation("store", true);

    info!(claim_id = %claim.id, "Claim created successfully");

    let response = ClaimResponse::from(&claim);
    Ok((StatusCode::CREATED, Json(response)))
}

/// Get claim by ID
#[utoipa::path(
    get,
    path = "/api/v1/claims/{id}",
    responses(
        (status = 200, description = "Claim found", body = ClaimResponse),
        (status = 404, description = "Claim not found", body = ApiError),
    ),
    params(
        ("id" = Uuid, Path, description = "Claim ID")
    ),
    tag = "claims"
)]
#[instrument(skip(state))]
pub async fn get_claim(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
) -> Result<Json<ClaimResponse>> {
    info!(claim_id = %id, "Retrieving claim");

    let claim = state
        .storage
        .retrieve(&id.to_string())
        .await
        .map_err(|_| ApiError::not_found("Claim"))?;

    Ok(Json(ClaimResponse::from(&claim)))
}

/// Add verification to claim
#[utoipa::path(
    put,
    path = "/api/v1/claims/{id}/verify",
    request_body = AddVerificationRequest,
    responses(
        (status = 200, description = "Verification added", body = ClaimResponse),
        (status = 404, description = "Claim not found", body = ApiError),
    ),
    params(
        ("id" = Uuid, Path, description = "Claim ID")
    ),
    tag = "claims"
)]
#[instrument(skip(state))]
pub async fn add_verification(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
    Json(req): Json<AddVerificationRequest>,
) -> Result<Json<ClaimResponse>> {
    info!(claim_id = %id, "Adding verification");

    // Retrieve claim
    let mut claim = state
        .storage
        .retrieve(&id.to_string())
        .await
        .map_err(|_| ApiError::not_found("Claim"))?;

    // Add verification
    let verification = Verification {
        verifier: req.verifier,
        signature: req.signature,
        timestamp: chrono::Utc::now(),
        notes: req.notes,
    };
    claim.add_verification(verification);

    // Update storage
    state.storage.store(&claim).await?;

    // Update query index
    state.query_engine.write().await.add_claim(&claim).await;

    info!(claim_id = %id, new_tier = ?claim.epistemic_tier, "Verification added");

    Ok(Json(ClaimResponse::from(&claim)))
}

/// Add provenance to claim
#[utoipa::path(
    put,
    path = "/api/v1/claims/{id}/provenance",
    request_body = AddProvenanceRequest,
    responses(
        (status = 200, description = "Provenance added", body = ClaimResponse),
        (status = 404, description = "Claim not found", body = ApiError),
    ),
    params(
        ("id" = Uuid, Path, description = "Claim ID")
    ),
    tag = "claims"
)]
#[instrument(skip(state))]
pub async fn add_provenance(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
    Json(req): Json<AddProvenanceRequest>,
) -> Result<Json<ClaimResponse>> {
    info!(claim_id = %id, "Adding provenance");

    // Retrieve claim
    let mut claim = state
        .storage
        .retrieve(&id.to_string())
        .await
        .map_err(|_| ApiError::not_found("Claim"))?;

    // Add provenance
    let mut prov = Provenance::new(req.source, req.source_type);
    if let Some(url) = req.url {
        prov = prov.with_url(url);
    }
    claim.add_provenance(prov);

    // Update storage
    state.storage.store(&claim).await?;

    info!(claim_id = %id, "Provenance added");

    Ok(Json(ClaimResponse::from(&claim)))
}
